core/commands/mfs/mfs.go

```diff
@@ -0,0 +1,689 @@
+
+			for _, o := range out.Entries {
+				if long {
+					fmt.Fprintf(buf, "%s\t%s\t%d\n", o.Name, o.Hash, o.Size)
+				} else {
+					fmt.Fprintf(buf, "%s\n", o.Name)
+				}
+			}
+			return buf, nil
+		},
+	},
+	Type: MfsLsOutput{},
+}
+
+var MfsPutCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "import a given file into a filesystem",
+		ShortDescription: `
+Mfs 'put' takes a file that already exists in ipfs and imports it into a
+given mfs filesystem.
+		`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("object", true, false, "ipfspath of object to import"),
+		cmds.StringArg("path", true, false, "path within filesystem to import to"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		rdir, err := getRootDirFromSession(node, req)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		obj := req.Arguments()[0]
+		objpath, err := path.ParsePath(obj)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		nd, err := node.Resolver.ResolvePath(req.Context(), objpath)
+		if err != nil {
+			res.SetError(fmt.Errorf("resolve path failed: %s", err), cmds.ErrNormal)
+			return
+		}
+
+		path := req.Arguments()[1]
+		err = mfs.PutNodeUnderDir(rdir, path, nd)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+	},
+}
+
+var MfsReadCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Read a file in a given mfs",
+		ShortDescription: `
+Read a specified number of bytes from a file at a given offset. By default, will
+read the entire file similar to unix cat.
+		`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to file to be read"),
+	},
+	Options: []cmds.Option{
+		cmds.IntOption("o", "offset", "offset to read from"),
+		cmds.IntOption("n", "count", "maximum number of bytes to read"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		rdir, err := getRootDir(req)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		path := req.Arguments()[0]
+		fsn, err := mfs.DirLookup(rdir, path)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		fi, ok := fsn.(*mfs.File)
+		if !ok {
+			res.SetError(fmt.Errorf("%s was not a file", path), cmds.ErrNormal)
+			return
+		}
+
+		offset, _, _ := req.Option("offset").Int()
+
+		_, err = fi.Seek(int64(offset), os.SEEK_SET)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		var r io.Reader = fi
+		count, found, err := req.Option("count").Int()
+		if err == nil && found {
+			r = io.LimitReader(fi, int64(count))
+		}
+
+		res.SetOutput(r)
+	},
+}
+
+var MfsMvCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Move files",
+		ShortDescription: `
+Move files around. Just like traditional unix mv.
+
+Example:
+
+    ipfs mfs -s myfs mv /a/b/c /foo/newc
```

Comment:
I like the idea of paths instead of a session var, although when dealing with the api, i've found the session flag to be easier. Less path manipulation to worry about.